integrations/GA4/browser.js

```diff
@@ -0,0 +1,211 @@
+/* eslint-disable class-methods-use-this */
+import logger from "../../utils/logUtil";
+import ScriptLoader from "../ScriptLoader";
+import {
+  eventName,
+  eventParameter,
+  itemParameter,
+} from "./ECommerceEventConfig";
+
+export default class GA4 {
+  constructor(config, analytics) {
+    this.measurementId = config.measurementId;
+    this.analytics = analytics;
+    this.sendUserId = config.sendUserId || false;
+    this.blockPageView = config.blockPageViewEvent || false;
+    this.name = "GA4";
+  }
+
+  loadScript(measurementId, userId) {
+    window.dataLayer = window.dataLayer || [];
+    window.gtag =
+      window.gtag ||
+      function gt() {
+        // eslint-disable-next-line prefer-rest-params
+        window.dataLayer.push(arguments);
+      };
+    window.gtag("js", new Date());
+
+    // This condition is not working, even after disabling page view
+    // page_view is even getting called on page load
+    if (this.blockPageView) {
+      window.gtag("config", measurementId, {
+        user_id: userId,
+        send_page_view: false,
+      });
+    } else {
+      window.gtag("config", measurementId);
+    }
+
+    ScriptLoader(
+      "google-analytics 4",
+      `https://www.googletagmanager.com/gtag/js?id=${measurementId}`
+    );
+  }
+
+  init() {
+    // To do :: check how custom dimension and metrics is used
+    const userId = this.analytics.userId || this.analytics.anonymousId;
+    this.loadScript(this.measurementId, userId);
+  }
+
+  // When adding events do everything ion lowercase.
+  // use underscores instead of spaces
+  // Register your parameters to show them up in UI even user_id
+
+  /* utility functions ---Start here ---  */
+  isLoaded() {
+    return !!(window.gtag && window.gtag.push !== Array.prototype.push);
+  }
+
+  isReady() {
+    return !!(window.gtag && window.gtag.push !== Array.prototype.push);
+  }
+  /* utility functions --- Ends here ---  */
+
+  isReservedName(eventName) {
+    const reservedName = [
+      "ad_activeview",
+      "ad_click",
+      "ad_exposure",
+      "ad_impression",
+      "ad_query",
+      "adunit_exposure",
+      "app_clear_data",
+      "app_install",
+      "app_update",
+      "app_remove",
+      "error",
+      "first_open",
+      "first_visit",
+      "in_app_purchase",
+      "notification_dismiss",
+      "notification_foreground",
+      "notification_open",
+      "notification_receive",
+      "os_update",
+      "screen_view",
+      "session_start",
+      "user_engagement",
+    ];
+
+    return reservedName.includes(eventName);
+  }
+
+  sendGAEvent(event, props) {
+    window.gtag("event", event, props);
+  }
+
+  getDestinationEvent(event) {
+    return eventName.find((p) => p.src.includes(event.toLowerCase()));
+  }
+
+  getDestinationEventProperties(props, destParameter) {
```

Comment:
same as above

Reply:
As method doc like what is the input and output format with an example

Reply:
better to use a library that does this? can you check once like get-value we use for transformer?

Reply:
done

Reply:
> better to use a library that does this? can you check once like get-value we use for transformer?

There is a library : " loadash " using that we can do, but loading some external library means loading extra js, isn't if we use our own implementation where we can modify code as required. which means we can move above function to global and make more generic. One of the way I can think is going through loadash code and provide same implementation in our utils.
What do you say, should I use external library or implement on our own .....

Reply:
we won't be loading a js but yes our total JS SDK size will increase. We can leave it
Can you add the examples for the input and output like
/*
props: {key: val, key2:val2}
destParameter: {..}
output: {...}
*/
getDestinationEventProperties(props, destParameter) {...}